streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1081,6 +1079,7 @@ private int getNumStreamThreads(final boolean hasGlobalTopology) {
                             } else {
                                 log.info("Successfully removed {} in {}ms", streamThread.getName(), time.milliseconds() - startMs);
                                 threads.remove(streamThread);
+                                queryableStoreProvider.removeStoreProvider(new StreamThreadStateStoreProvider(streamThread));
```

Comment:
The test actually passes with this line commented out, this is because in StreamThreadStateStoreProvider it deals with having dead threads left in there - https://github.com/apache/kafka/blob/4fd6d2bec8a6249e211b9769e887cc24d5cf1444/streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java#L47. So without this line this if condition gets hit. With this `removeStoreProvider` this condition isn't hit, so it's a small optimisation.

Reply:
Ah, yeah I guess it would have always had to handle `DEAD` threads since in the before-time (ie before we added the `add/removeStreamThread` APIs) it was always possible for a thread to just die when hit with an unexpected exception.

That said, I feel a lot better about trimming a removed thread from the list explicitly. Don't want to build up a mass grave of mostly dead threads (well, dead thread store providers) that can never be garbage collected over the application's lifetime

Reply:
It seems a bit weird to have to create a `new StreamThreadStateStoreProvider(streamThread)` just to remove and existing `StreamThreadStateStoreProvider` from this -- can we maybe change it so that `#removeStoreProvider` accepts a `StreamThread` reference, or even just a `String streamThreadName`? And then store a map from name to `StreamThreadStateStoreProvider` or something -- WDYT?

Reply:
Yeah agreed - that's a better idea, I've changed to a Map and being able to remove by thread name. To be honest, I'd like to just pass in StreamThread into the #addStoreProvider but then it's hard to test because the StreamThreadStateStoreProvider is instantiated inside #addStoreProvider rather than being able to inject a stub (this would also allow #removeStoreProvider to just be passed a StreamThread which would make it more consistent).